### PR TITLE
ENH: Add overflow handling for negative integers scalar multiplication 

### PR DESCRIPTION
**Pull request by Micky774**
## Related Issues/PRs
Related to https://github.com/numpy/numpy/issues/21506

Follow-up to https://github.com/numpy/numpy/pull/21648

## Changes
Adds checks for multiplication overflow for signed integers, in particular when they are negative.

## Comments
N/A

### CODE REVIEW

**Comment by seberg** [comment]
I am wondering slightly whether we should care that `npy_mul_with_overflow_intp` is currently almost only used in places where we really don't care about negative values, because its main purpose is/was for shape calculations.
Likely, it just doesn't matter, though, and this is good.

**Comment by Micky774** [comment]
> I am wondering slightly whether we should care that `npy_mul_with_overflow_intp` is currently almost only used in places where we really don't care about negative values, because its main purpose is/was for shape calculations. Likely, it just doesn't matter, though, and this is good.

I currently have it treated as an "unsigned" value for that exact reason. It doesn't break any tests so hopefully it isn't problematic.

**Comment by seberg** [comment]
@Micky774 did you prefer keeping it as is, or rather introduce a new name for the `intp` version?  I am slightly leaning to that right now.  (even if it is a bit annoying code wise.)

**Comment by Micky774** [comment]
> @Micky774 did you prefer keeping it as is, or rather introduce a new name for the `intp` version? I am slightly leaning to that right now. (even if it is a bit annoying code wise.)

I can adopt your suggestion and write a separate function for it. Apologies for the delays, I usually work on numpy stuff towards the end of the week :)

File: numpy/core/src/common/templ_common.h.src [hunk 1]
```diff
@@ -34,13 +36,41 @@ npy_mul_with_overflow_@name@(@type@ * r, @type@ a, @type@ b)
...
 /**end repeat**/
 
+static NPY_INLINE int
+npy_mul_sizes_with_overflow (npy_intp * r, npy_intp a, npy_intp b)
+{
+#ifdef HAVE___BUILTIN_MUL_OVERFLOW
+    return __builtin_mul_overflow(a, b, r);
+#else
+
+    assert a >= 0 && b >= 0, "this function only supports non-negative numbers"
```
Comment (seberg) [inline numpy/core/src/common/templ_common.h.src:61]:
```suggestion
    assert(a >= 0 && b >= 0, "this function only supports non-negative numbers")
```
I didn't realize that works... but it seems like it doesn't on MSVC.

**Comment by Micky774** [comment]
@seberg I think this is ready now, there's just a problem w/ Travis CI

**Comment by seberg** [comment]
Agreed, thanks @Micky774!

**Pull request merged**